packages/radix-ui-themes/src/helpers/responsive.ts

```diff
@@ -0,0 +1,181 @@
+          const baseClassName = getBaseClassName(className, value, parseValue);
+          const bpClassName = bp === 'initial' ? baseClassName : `${bp}:${baseClassName}`;
+          classNames.push(bpClassName);
+        } else if (allowArbitraryValues) {
+          const bpClassName = bp === 'initial' ? className : `${bp}:${className}`;
+          classNames.push(bpClassName);
+        }
+      }
+    }
+
+    return classNames.join(' ');
+  }
+
+  if (allowArbitraryValues) {
+    return className;
+  }
+}
+
+function getBaseClassName(
+  className: string,
+  value: string,
+  parseValue: (value: string) => string | undefined
+): string {
+  const delimiter = className ? '-' : '';
+  const matchedValue = parseValue(value);
+  const isNegative = matchedValue?.startsWith('-');
+  const minus = isNegative ? '-' : '';
+  const absoluteValue = isNegative ? matchedValue?.substring(1) : matchedValue;
+  return `${minus}${className}${delimiter}${absoluteValue}`;
+}
+
+interface GetResponsiveCustomPropertiesOptions {
+  customProperty: `--${string}`;
+  value: Responsive<StringOrValue<string>> | Responsive<string> | undefined;
+  propValues: string[] | readonly string[];
+  parseValue?: (value: string) => string | undefined;
+}
+
+function getResponsiveCustomProperties({
+  customProperty,
+  value,
+  propValues,
+  parseValue = (value) => value,
+}: GetResponsiveCustomPropertiesOptions) {
+  let styles: Record<string, string | undefined> = {};
+
+  // Don't generate custom properties if the value is not arbitrary
+  if (!value || (typeof value === 'string' && propValues.includes(value))) {
+    return undefined;
+  }
+
+  if (typeof value === 'string') {
+    const customProperties = getCustomProperties(customProperty);
+    styles = Object.fromEntries(customProperties.map((prop) => [prop, value]));
+  }
+
+  if (isResponsiveObject(value)) {
+    const object = value;
+
+    for (const bp in object) {
+      // Make sure we are not iterating over keys that aren't breakpoints
+      if (!hasOwnProperty(object, bp) || !breakpoints.includes(bp)) {
+        continue;
+      }
+
+      const value = object[bp];
+      const customProperties = getCustomProperties(customProperty);
+
+      // Don't generate a custom property if the value is not arbitrary
+      if (propValues.includes(value)) {
+        continue;
+      }
+
+      for (const customProperty of customProperties) {
+        const bpProperty = bp === 'initial' ? customProperty : `${customProperty}-${bp}`;
+
+        styles = {
+          [bpProperty]: value,
+          ...styles,
+        };
+      }
+    }
+  }
+
+  for (const key in styles) {
+    const value = styles[key];
+    if (value !== undefined) {
+      styles[key] = parseValue(value);
+    }
+  }
+
+  return styles;
+}
+
+// Split comma-separated custom properties. This is an escape hatch for
+// when you need to generate multiple custom properties with the same value.
+function getCustomProperties(str: `--${string}`) {
```

Comment:
Should we be more explicit and accept this?
```ts
`--${string}` | `--${string}`[]
```

so instead of:

```ts
customProperty: '--margin-left, --margin-right',
```

we pass:

```ts
customProperty: ['--margin-left', '--margin-right'],
```

Reply:
I feel like we should yes, because that's clearly the intent here that it can have single or multiple custom properties, so using string with a separator to split on feels like a hack we're shielding away from the compiler.

Reply:
Updated to
```ts
customProperties: `--${string}`[]
```

The middle ground was fancy-ish, but not fancy enough to keep